samples/nrf9160/modem_trace_flash/src/trace_flash.c

```diff
@@ -0,0 +1,188 @@
+static trace_backend_processed_cb trace_processed_callback;
+
+static int internal_flash_write(const void *data, size_t len, bool flush)
+{
+	int64_t uptime_ref = k_uptime_get();
+	int64_t write_time;
+
+	int err = stream_flash_buffered_write(&stream, data, len, flush);
+
+	if (err != 0) {
+		LOG_ERR("stream_flash_buffered_write error %d", err);
+		return err;
+	}
+
+	write_time = k_uptime_delta(&uptime_ref);
+	tot_write_time += write_time;
+
+	traces_size += len;
+
+	return (int)len;
+}
+
+static const uint32_t magic_val = 0x152ac523;
+int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
+{
+	int err;
+
+	LOG_DBG("trace_backend_init\n");
+	if (trace_processed_cb == NULL) {
+		return -EFAULT;
+	}
+
+	trace_processed_callback = trace_processed_cb;
+
+	if (magic != magic_val) {
+		LOG_DBG("Initializing .noinit\n");
+		magic = magic_val;
+		traces_size = 0;
+	}
+
+	traces_size = 0;
+
+	err = flash_area_open(FLASH_AREA_ID(modem_trace), &modem_trace_area);
+	if (err) {
+		LOG_ERR("flash_area_open error:  %d", err);
+		return -ENODEV;
+	}
+
+	err = flash_area_has_driver(modem_trace_area);
+	if (err != 1) {
+		LOG_ERR("flash_area_has_driver error: %d\n", err);
+	}
+
+	flash_dev = flash_area_get_device(modem_trace_area);
+	if (flash_dev == NULL) {
+		LOG_ERR("Failed to get flash device\n");
+		return -ENODEV;
+	}
+
+	err = stream_flash_init(&stream, flash_dev, stream_buf, BUF_LEN, TRACE_OFFSET, TRACE_SIZE,
+		NULL);
+	if (err) {
+		LOG_ERR("stream_flash_init error: %d", err);
+		return err;
+	}
+
+	int64_t erase_before = k_uptime_get();
+
+	/* In order to write modem traces uninterrupted at full speed, we need to erase the flash
+	 * before starting the modem tracing.
+	 */
+	err = flash_erase(flash_dev, TRACE_OFFSET, TRACE_SIZE);
+	if (err) {
+		LOG_ERR("flash_erase error: %d", err);
+		return err;
+	}
+	int64_t erase_time = k_uptime_delta(&erase_before);
+
+	LOG_DBG("Erased %lu bytes in %lld ms", (unsigned long) TRACE_SIZE, erase_time);
+	LOG_DBG("Erase speed %lld kb/s ", (TRACE_SIZE * 8 / erase_time));
+
+	tot_write_time = 0;
+
+	LOG_DBG("Modem trace flash backend initialized\n");
+
+	return 0;
+}
+
+int trace_backend_deinit(void)
+{
+	LOG_DBG("Flash trace backend deinitialized\n");
+
+	return 0;
+}
+
+int trace_backend_write(const void *data, size_t len)
+{
+	if (has_flushed) {
+		trace_processed_callback(len);
+		return -EIO;
+	}
+
+	int ret = internal_flash_write(data, len, false);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = trace_processed_callback(len);
+	if (ret < 0) {
+		LOG_ERR("trace_processed_callback failed: %d", ret);
+		return ret;
+	}
+
+	return (int)len;
+}
+
+int trace_flash_flush(void)
+{
+	int err = internal_flash_write(NULL, 0, true);
+
+	if (err != 0) {
+		LOG_ERR("Flush to flash failed: %d", err);
+		return err;
+	}
+
+	has_flushed = true;
+
+	LOG_DBG("Written %d bytes in %lld ms", traces_size, tot_write_time);
+	LOG_DBG("Throughput = %lld kb/s ", (traces_size * 8 / tot_write_time));
```

Comment:
Thanks, I wasn't aware of that. Tried it now, but I prefer the average speed for all trace writes instead of printing periodically. Perhaps possible to configure, but have to look into that later.